source/geod24/concurrency.d

```diff
@@ -1006,6 +1007,601 @@ protected:
+                return true;
+            }
+            return false;
+        }
+
+        /***********************************************************************
+
+            Returns if SelectState is already consumed or not
+
+        ***********************************************************************/
+
+        bool isConsumed()
+        {
+            return atomicLoad(this.consumed);
+        }
+
+        /// ID of the select call that consumed this SelectState
+        uint id;
+
+        /// Success/failure state of the select call with id of `id`
+        bool success;
+
+    private:
+        /// Indicates if this SelectState is consumed or not
+        shared(bool) consumed;
+    }
+
+    /***********************************************************************
+
+        A golang style channel implementation with buffered and unbuffered
+        operation modes
+
+        Intended to be used between Fibers
+
+    ***********************************************************************/
+
+    class Channel (T)
+    {
+
+        /***********************************************************************
+
+            Constructs a Channel
+
+            Param:
+                max_size = Maximum amount of T a Channel can buffer
+                                0 -> Unbufferd operation
+                                Positive integer -> Buffered operation
+
+        ***********************************************************************/
+
+        this(ulong max_size = 0)
+        {
+            this.max_size = max_size;
+            this.lock = new Mutex;
+        }
+
+        /***********************************************************************
+
+            Write a message to the Channel
+
+            Unbuffered mode:
+
+                If a reader is already blocked on the Channel, writer copies the
+                message to reader's buffer and wakes up the reader by yielding
+
+                If no reader is ready in the wait queue, writer appends itself
+                to write wait queue and blocks
+
+            Buffered mode:
+
+                If a reader is already blocked on the Channel, writer copies the
+                message to reader's buffer and wakes up the reader by yielding
+
+                If no reader is in the wait queue and the buffer is not full
+                writer puts the message in the Channel buffer and returns immediatly
+
+                If no reader is in the wait queue and the buffer is full
+                writer appends itself to write wait queue and blocks
+
+            If channel is closed, it returns immediatly with a failure,
+            regardless of the operation mode
+
+            Param:
+                val = Message to write to channel
+            Returns:
+                Success/Failure
+
+        ***********************************************************************/
+
+        bool write () (auto ref T val)
+        {
```

Comment:
`tryWrite` acquires the lock. It releases the lock if it successfully completes a non-blocking operation.
In case of a blocking operation; `tryWrite` returns with the mutex still locked and `enqueueEntry` unlocks when it is finished.

Reply:
Ah yes. That makes sense, thanks!